### PR TITLE
cmd: tx send command sends transaction to network

### PR DESCRIPTION
**Pull request by Ali-Meh**
## Description
seprate `tx send` from `payload send` in cmd for better UX

- Seprate sendtx and payload sendtx in cmd
- add  stamp to blockchian info stamp and update proto
- confirm to publish tx
- gRPC transaction function error made verbose
- Seprate payload and tx in cli
- Add gRPC Client in cmd
- Confirm to publish bondtx

## Related issue(s)
#185
#195

## Checklist

- [x] New feature (non-breaking change which adds functionality)
- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] CHANGELOG is updated.

### CODE REVIEW

File: cmd/gRPC.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
+package cmd
```
Comment (b00f) [inline cmd/gRPC.go:1]: For the name of file is better to keep it all lowercase: "grpc.go"

File: cmd/zarb/main.go [hunk 2]
```diff
@@ -21,9 +22,13 @@ func zarb() *cli.Cli {
...
 		k.Command("change-auth", "Change the passphrase of a keyfile", key.ChangeAuth())
 	})
-	app.Command("tx", "Create raw transaction", func(k *cli.Cmd) {
-		k.Command("bond", "Generate raw bond transaction", tx.BondTx())
-		k.Command("send", "Generate raw send transaction", tx.SendTx())
+	app.Command("tx", "Create, Sing, Publish transactions", func(k *cli.Cmd) {
+		k.Command("bond", "Send signed Bond transaction to network", tx.BondTx())
+		k.Command("send", "Send signed Send transaction to network", tx.SendTx())
+	})
+	app.Command("payload", "Create raw transaction", func(k *cli.Cmd) {
```
Comment (b00f) [inline cmd/zarb/main.go:29]: AS I checked your code, I am convinced that maybe we don't need the old style.

File: www/grpc/proto/zarb.proto [hunk 3]
```diff
@@ -75,6 +75,7 @@ message BlockchainInfoRequest {
 
 message BlockchainInfoResponse {
   int64 height=1;
+  string stamp=2;
```
Comment (b00f) [inline www/grpc/proto/zarb.proto:78]: Use `Last_block_hash` instead. 
Stamp is not a blockchain thing. Stamp is a Tx related thing.

**Review by b00f — CHANGES_REQUESTED** [review]
@d4rk5h0t 
Thanks for this PR. Sounds good. I think we might don't need `zarb tx payload` subcommand anymore. Let me check it. 

Another idea, maybe we can relocate `grpc.go` somewhere else like `util` folder. Because we can use this inside MainTest or other purposes later. Keep it in mind. Later we can discuss about it.

File: cmd/zarb/main.go [hunk 4]
```diff
@@ -21,9 +22,13 @@ func zarb() *cli.Cli {
 		k.Command("verify", "Verify a signature", key.Verify())
 		k.Command("change-auth", "Change the passphrase of a keyfile", key.ChangeAuth())
 	})
-	app.Command("tx", "Create raw transaction", func(k *cli.Cmd) {
-		k.Command("bond", "Generate raw bond transaction", tx.BondTx())
-		k.Command("send", "Generate raw send transaction", tx.SendTx())
+	app.Command("tx", "Create, Sing, Publish transactions", func(k *cli.Cmd) {
```
Comment (b00f) [inline cmd/zarb/main.go:25]: `Create, sign and publish a transaction` is better. 
sign or sing ;)

File: cmd/zarb/main.go [hunk 5]
```diff
@@ -21,9 +22,13 @@ func zarb() *cli.Cli {
...
+		k.Command("bond", "Send signed Bond transaction to network", tx.BondTx())
```
Comment (b00f) [inline cmd/zarb/main.go:26]: `Create, sign and publish a bond transaction`, may be better

File: cmd/zarb/main.go [hunk 6]
```diff
@@ -21,9 +22,13 @@ func zarb() *cli.Cli {
...
+		k.Command("send", "Send signed Send transaction to network", tx.SendTx())
```
Comment (b00f) [inline cmd/zarb/main.go:27]: `Create, sign and publish a send transaction`, may be better

File: cmd/zarb/payload/send_tx.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+			// ---
+			if *seqOpt == 0 {
+				cmd.PrintWarnMsg("Sequence number is not defined.")
+				c.PrintHelp()
+				return
+			}
+			seq = *seqOpt
+
+			if *amountOpt == 0 {
+				cmd.PrintWarnMsg("Stake is not defined.")
```
Comment (b00f) [inline cmd/zarb/payload/send_tx.go:69]: "Amount is not defined."

**Comment by b00f** [comment]
I think we can remove `fee` option. Only show the last amount+fee that user is going to pay.

File: cmd/zarb/tx/common.go [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+	k, err := key.DecryptKeyFile(keyfile, auth)
+	if err != nil {
+		cmd.PrintErrorMsg("Couldn't retrieve Key: %v", err)
+		return
+	}
+	k.ToSigner().SignMsg(trx)
+
+	cmd.PrintWarnMsg("you are about to publish:")
+	cmd.PrintJSONObject(trx)
+	confirm := cmd.PromptInput("are you sure [y/n]:")
```
Comment (b00f) [inline cmd/zarb/tx/common.go:23]:
```
confirm := cmd.PromptInput("The operation is not reversible. Are you sure? [yes/no]")
	if !strings.HasPrefix(strings.ToLower(confirm), "yes") {
```

Comment (b00f) [reply cmd/zarb/tx/common.go:23]: This is really important. Let user write `yes`. Make sure not accidentally press a `y` key.

File: cmd/zarb/tx/common.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/zarbchain/zarb-go/keystore/key"
+	"github.com/zarbchain/zarb-go/tx"
+	zarb "github.com/zarbchain/zarb-go/www/grpc/proto"
+)
+
+func signAndPublish(trx *tx.Tx, keyfile, auth string, grpcClient zarb.ZarbClient) {
+	//sign transaction
+	k, err := key.DecryptKeyFile(keyfile, auth)
+	if err != nil {
+		cmd.PrintErrorMsg("Couldn't retrieve Key: %v", err)
```
Comment (b00f) [inline cmd/zarb/tx/common.go:16]: `Couldn't retrieve the key`

File: cmd/zarb/tx/common.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+		return
+	}
+	k.ToSigner().SignMsg(trx)
+
+	cmd.PrintWarnMsg("you are about to publish:")
+	cmd.PrintJSONObject(trx)
```
Comment (b00f) [inline cmd/zarb/tx/common.go:22]: better print the raw bytes for user that want to send the raw transaction.

use `cmd.PrintLine()` for line separator.

File: cmd/zarb/tx/bond_tx.go [hunk 11]
```diff
@@ -112,10 +110,52 @@ func BondTx() func(c *cli.Cmd) {
...
+
+			grpcClient, err := cmd.GetRPCClient(rpc)
+			if err != nil {
+				cmd.PrintErrorMsg("Couldn't connect to RPC Server: %v", err)
+				return
+			}
+
+			seq = *seqOpt
+			if *seqOpt == 0 {
+				seq, err = cmd.GetSequence(grpcClient, bonder)
```
Comment (b00f) [inline cmd/zarb/tx/bond_tx.go:140]: I have a suggestion here. 
If you make `func grpcClient(grpcOpt *string) {... }`
Then you only ask user anytime you want to query or broadcast the network. 
`cmd.GetSequence(grpcClient(grpcOpt), bonder)`

So if user wants to set all the options manually (offline), no need to ask gRPC endpoint.

File: cmd/zarb/tx/common.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+		cmd.PrintWarnMsg("Opration aborted!")
+		return
+	}
+
+	// publish
+	signedTrx, _ := trx.Encode()
+	if id, err := cmd.SendTx(grpcClient, signedTrx); err != nil {
+		cmd.PrintErrorMsg("Couldn't publish transaction: %v", err)
+	} else {
+		cmd.PrintSuccessMsg("transaction sent with Id: %v", id)
```
Comment (b00f) [inline cmd/zarb/tx/common.go:34]: "Transaction sent with ID"

**Comment by b00f** [comment]
I wish we could write test for this part. because it is really important. Later we can discuss about how to write test for CLI.

**Comment by d4rk5h0t** [comment]
> 
> 
> @d4rk5h0t
> Thanks for this PR. Sounds good. I think we might don't need `zarb tx payload` subcommand anymore. Let me check it.
> 
> Another idea, maybe we can relocate `grpc.go` somewhere else like `util` folder. Because we can use this inside MainTest or other purposes later. Keep it in mind. Later we can discuss about it.

sure but it wasn't me 😂

File: cmd/gRPC.go (hunk 1 above)
Comment (Ali-Meh) [reply cmd/gRPC.go:1]: sure moved it to util/grpc.go

File: cmd/zarb/main.go (hunk 2 above)
Comment (Ali-Meh) [reply cmd/zarb/main.go:29]: we might need them for smart-contracts implementations so i suggest keeping them

File: www/grpc/proto/zarb.proto (hunk 3 above)
Comment (Ali-Meh) [reply www/grpc/proto/zarb.proto:78]: sure applyed

File: cmd/zarb/main.go (hunk 4 above)
Comment (Ali-Meh) [reply cmd/zarb/main.go:25]: you right my bad

File: cmd/zarb/main.go (hunk 5 above)
Comment (Ali-Meh) [reply cmd/zarb/main.go:26]: done

**Comment by Ali-Meh** [comment]
> I wish we could write test for this part. because it is really important. Later we can discuss about how to write test for CLI.

@b00f 
I agree with you, Actually I tried to implement some tests for CLI but I realized it's heavily relies on tested modules like gRPC or transaction so may be pushing it much.
besides it needs some mocks to implement if we could use some libraries to generate mocks, it would be much easier but it may have inconsistency with current code base

File: cmd/zarb/main.go (hunk 2 above)
Comment (b00f) [reply cmd/zarb/main.go:29]: This is redundant to me now. With the new commands, we can still make transactions and sign them off line. 
Please consider that users might raise issues about them. Let archive them now.

File: cmd/zarb/tx/bond_tx.go (hunk 11 above)
Comment (d4rk5h0t) [reply cmd/zarb/tx/bond_tx.go]: @b00f
I tried to implement it this idea.
let me know if you got better way

Comment (Ali-Meh) [reply cmd/zarb/tx/bond_tx.go]: i tried to implement this idea let me know if you got better way

File: cmd/zarb/tx/bond_tx.go [hunk 13]
```diff
@@ -112,10 +110,40 @@ func BondTx() func(c *cli.Cmd) {
...
+			}
+
+			stamp, err = crypto.HashFromString(*stampOpt)
+			if err != nil {
+				stamp, err = util.GetStamp(promptRPCEndpoint(*grpcOpt))
+				if err != nil {
+					cmd.PrintErrorMsg("Couldn't retrieve stamp from RPC Server: %v", err)
+					return
+				}
+			}
```
Comment (b00f) [inline cmd/zarb/tx/bond_tx.go:142]: This part has a potential problem. User may enter invalid hash. So we ignore it and query from the node. 
Maybe we can rewrite it like this:
```
			
			if stampOpt == nil {
                stamp, err = crypto.HashFromString(*stampOpt)
                if err != nil {
					cmd.PrintErrorMsg("Stamp is not valid: %v", err)
					return
				}
           } else { 
				stamp, err = util.GetStamp(promptRPCEndpoint(*grpcOpt))
				if err != nil {
					cmd.PrintErrorMsg("Couldn't retrieve stamp from RPC Server: %v", err)
					return
				}
			}
```

File: cmd/zarb/tx/bond_tx.go [hunk 14]
```diff
@@ -112,10 +110,40 @@ func BondTx() func(c *cli.Cmd) {
...
+
+			//RPC
+			seq = *seqOpt
+			if *seqOpt == 0 {
+				seq, err = util.GetSequence(promptRPCEndpoint(*grpcOpt), bonder)
+				if err != nil {
+					cmd.PrintErrorMsg("Couldn't retrieve Sequence number from RPC Server: %v", err)
+					return
+				}
+			}
```
Comment (b00f) [inline cmd/zarb/tx/bond_tx.go:133]: How about this:
```
        if seqOpt != nil {
                seq = *seqOpt
           } else { 
				seq, err = util.GetSequence(promptRPCEndpoint(*grpcOpt), bonder)
				if err != nil {
					cmd.PrintErrorMsg("Couldn't retrieve Sequence number from RPC Server: %v", err)
					return
				}
			}
```

File: account/account.go [hunk 15]
```diff
@@ -72,8 +72,9 @@ func (acc *Account) UnmarshalJSON(bs []byte) error {
 }
 
 func (acc Account) Fingerprint() string {
-	return fmt.Sprintf("{ %s %v}",
+	return fmt.Sprintf("{ %s %d %v}",
```
Comment (b00f) [inline account/account.go:75]: Could you please update Fingerprint for validator as well?

Comment (Ali-Meh) [reply account/account.go:75]: sure why not

**Review by b00f — APPROVED** [review]

**Pull request merged**